source/dub/project.d

```diff
@@ -1957,3 +1979,39 @@ private immutable DefaultTestRunnerCode = q{
 		}
 	}
 };
+
+
+struct SelverLookupResult {
+	NativePath absolutePath; // to dub.selections.json
+	SelectedVersions selectedVersions;
+}
+
+// Does both lookup *and* parsing because a parent dir's dub.selections.json
+// file is only inherited if it has `"inheritable": true` (=> needs parsing).
+Nullable!SelverLookupResult lookupAndParseSelectionsFile(NativePath absRootPackagePath)
+	in(absRootPackagePath.absolute)
+{
+	alias N = typeof(return);
+
+	// check for dub.selections.json in root package dir first, then walk up its
+	// parent directories
+	for (NativePath dir = absRootPackagePath; dir.hasParentPath; dir = dir.parentPath) {
+		const selverfile = dir ~ SelectedVersions.defaultFile;
+		if (existsFile(selverfile)) {
+			// TODO: Remove `StrictMode.Warn` after v1.40 release
+			// The default is to error, but as the previous parser wasn't
+			// complaining, we should first warn the user.
+			auto selected = parseConfigFileSimple!Selected(selverfile.toNativeString(), StrictMode.Warn);
+			const isValid = !selected.isNull() && (dir == absRootPackagePath || selected.get().inheritable);
+			if (isValid) {
+				return N(SelverLookupResult(
+					selverfile,
+					new SelectedVersions(selected.get(), dir.relativeTo(absRootPackagePath))
+				));
+			}
```

Comment:
in case of invalid selections files, shouldn't we rather error than silently ignoring the file?

Reply:
I'm pretty sure I kept the behavior as-is (at that time - merge conflicts now obviously after all that time) - an empty `Selections` object if there's no file or the file is invalid.